### ISSUE TITLE
Can't connect to Cloud SQL Postgres using socket

### ISSUE THREAD

**Issue by Verhaeg** [prisma/prisma#11171]
### Bug description

According to some other issues, ([#1508](https://github.com/prisma/prisma/issues/1508)) and ([#8638](https://github.com/prisma/prisma/issues/8638)) there was some issue when trying to connect with sockets and it was changed at least for MySQL.

In our case we use CloudSQL Postgres and configuring Cloud Run to connect to this instance using Cloud Proxy where it creates a socket usually in `/cloudsql` folder by creating a new folder inside with the resource link in the format of `<project>:<region>:<instance>` and then a file inside `.s.PGSQL.5432`. 

We set the `DATABASE_URL` following all rules, something in the lines of:
```text
postgresql://<USER>:<PASSWORD>@localhost/<DATABASE>?host=/cloudsql/<CLOUDSQL_INSTANCE_RESOURCE_NAME>&schema=public
```

We tried several combinations but without any success and the error we found was:
```text
Can't reach database server at `/cloudsql/<CLOUDSQL_INSTANCE_RESOURCE_NAME>`:`5432`
```

Looking at the code it seems to be forcing a TLS connection instead of allowing socket like it does for MySQL.

For now the workaround we are working with is (For anyone facing this issue):
1. Enabling Private IP in our CloudSQL instances
2. Created a Serverless VPC Connector (with extra charges =/ )
3. Attaching our Cloud Run service to this VPC connector and using the private IP

### How to reproduce

Configure any socket connection for Postgres and try to connect using this socket.

### Expected behavior

Be able to connect to a Postgres instance through UNIX socket connection.

### Prisma information

prisma:client clientVersion: 2.29.1

### Environment & setup

- Database: Postgres
- Node.js version: 14.17.5

### Prisma Version

```
2.29.1
```

**Comment by rained23** [comment]
having the same issue, but in our production env it is working.

Prisma version 3.6.0

**Comment by yoshitamagoshi** [comment]
I was able to get it to work with the following `DATABASE_URL` format:

`postgresql://<DB_USER>:<DB_PASS>@localhost/<DB_NAME>?host=/cloudsql/<CLOUD_SQL_CONNECTION_NAME>`

Ensure that `CLOUD_SQL_CONNECTION_NAME` is of the format `project_id:region:instance_name` and not just the instance name.

Also ensure that in the Cloud Run Connections settings for the revision that you've enabled the appropriate Cloud SQL Connection which will create the cloud sql proxy connection for you as per the [instructions here](https://cloud.google.com/sql/docs/mysql/connect-run?authuser=1&_ga=2.206344090.-886907616.1642687419).

Credit to [this video](https://www.youtube.com/watch?v=WLsFF4mtqXQ) for pointing me in the right direction.

**Comment by Verhaeg** [comment]
Hi @yoshitamagoshi, we did try this setting, the only difference to our approach is that we need to set the schema as well. And we did URI encode as requested as well.

This is the URL I'm using (with the proper changes) and we are still receiving the same error:

`postgresql://<USER>:<PASSWORD>@localhost/<DATABASE>?host=/cloudsql/<CONNECTION_INSTANCE_NAME>&schema=public`


```
\nInvalid `prisma.customer.findMany()` invocation:\n\n\n  Can't reach database server at `/cloudsql/<CLOUD_INSTANCE_NAME>`:`5432`\n\nPlease make sure your database server is running at `/cloudsql/<CLOUD_INSTANCE_NAME>`:`5432`.
```

**Comment by yoshitamagoshi** [comment]
@Verhaeg Does the service account you're using for that Cloud Run revision have the `Cloud SQL Client` permission?

**Comment by Verhaeg** [comment]
@yoshitamagoshi Yes.. that was missing.. Due to the error message I didn't think about permissions =/

Thanks for pointing it out!

**Comment by janpio** [comment]
I think this is something we saw a few times in the past already. Is this something we could surface better in Prisma?

**Comment by rained23** [comment]
for my case I am using Nest.js + Prisma, I am not sure it is caused by Prisma or Cloud Run, it only affects some modules.
So when I try to call my API that uses Prisma, it throws the `Invalid invocation`  but it works on other APIs. 

So I can say that the error thrown is not really valid because Prisma actually can connect to the DB for a different tables.

**Comment by irwan-iprospect** [comment]
i got same issue here. i already test with different project to just install next js and prisma, the result is no issue. and then after i try to adding and update some library, i got this issue.

`{
  "name": "test-project",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@fortawesome/fontawesome-svg-core": "^1.2.35",
    "@fortawesome/free-brands-svg-icons": "^5.15.3",
    "@fortawesome/free-solid-svg-icons": "^5.15.3",
    "@fortawesome/react-fontawesome": "^0.1.14",
    "@next-auth/prisma-adapter": "^0.5.4",
    "@next-tools/dynamic": "^0.0.11",
    "prisma": "^3.9.0",
    "@types/react": "^17.0.25",
    "babel-plugin-superjson-next": "^0.4.1",
    "eslint": "7.31.0",
    "eslint-config-next": "11.0.1",
    "fs": "^0.0.1-security",
    "image-size": "^1.0.1",
    "multiparty": "^4.2.2",
    "next": "^11.1.2",
    "next-auth": "^3.29.0",
    "next-connect": "^0.10.2",
    "pptxgenjs": "^3.9.0",
    "react": "17.0.2",
    "react-bootstrap-sweetalert": "^5.2.0",
    "react-data-table-component": "^7.4.4",
    "react-dom": "17.0.2",
    "react-hook-form": "^7.22.5",
    "react-sortablejs": "^6.0.0",
    "sortablejs": "^1.14.0",
    "styled-components": "^5.3.3",
    "superjson": "^1.7.5",
    "sweetalert2": "^11.1.7",
    "typescript": "^4.4.3"
  },
  "devDependencies": {
    "@types/bcrypt": "^5.0.0",
    "@types/bcrypt-nodejs": "^0.0.31",
    "@types/formidable": "^2.0.4",
    "@types/fs-extra": "^9.0.13",
    "@types/styled-components": "^5.1.21",
    "bcrypt": "^5.0.1",
    "bcrypt-nodejs": "^0.0.3",
    "formidable": "^1.2.2",
    "fs-extra": "^10.0.0"
  }
}
`

**Comment by janpio** [comment]
Please open a new issue @irwan-iprospect and provide all the information the bug issue template is asking for. Thanks.

**Comment by Kukunin** [comment]
For anybody who encounters the same issue with Cloud Run and verified the SQL Connection and IAM permissions, make sure that you don't have background jobs and "CPU is only allocated during request processing" setting. It was my case, the service didn't have enough CPU resources so connections were dropped. 

I converted to do all logic within the request scope, and it worked fine

**Comment by dustinlyons** [comment]
@Kukunin THANK YOU! I spent hours troubleshooting why Google Cloud Run + Cloud SQL could not fetch a connection from the connection pool. 

```
PrismaClientInitializationError:

Invalid `prisma.checkPoint.findFirst()` invocation:

Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 3)

Container called exit(1).
```

The "CPU is only allocated during request processing" setting you mentioned was the solution. I think this is easy to miss if you don't spend too much time in their UI (which I don't, using terraform).

My application was not a traditional request/response API and thus needed CPU available all the time to use Prisma and the database.

**Comment by otopba** [comment]
> "CPU is only allocated during request processing"

Thank you so much!

**Comment by shanirosen-chram** [comment]
Hey, anyone got this working with IAM instead of user/password? They only way I got it working is by using google cloud connector with local proxy

**Comment by Verhaeg** [comment]
Hi @shanirosen-chram not sure if a limitation from Prisma connection right now, but we came back to another project using Prisma + CloudRun + CloudSQL and we are trying to connect using IAM (Service Account).

Locally, I was able to connect using IAM with a little "combo":

1. Started `cloud-sql-proxy` using the desired SA to create the socket. This will "mimic" CloudRun connection with CloudSQL
2. Using `@google-cloud/cloud-sql-connector` started a localProxy - similar to https://github.com/GoogleCloudPlatform/cloud-sql-nodejs-connector/blob/main/examples/prisma/postgresql/connect.ts

```js
import { Connector, AuthTypes, IpAddressTypes } from '@google-cloud/cloud-sql-connector'
import { PrismaClient } from '@prisma/client'

//...

const connector = new Connector()
connector.startLocalProxy({
  instanceConnectionName,
  ipType: IpAddressTypes.PUBLIC,
  authType: AuthTypes.IAM,
  listenOptions: { path: '.s.PGSQL.5432' },
})
const hostPath = process.cwd()
const datasourceUrl = `postgresql://${POSTGRES_USER}@localhost/${POSTGRES_DB}?host=${hostPath}`

this.client = new PrismaClient({
  datasourceUrl,
})
```

We still have to test in CloudRun.. probably start the localProxy as a standalone daemon before starting the application and have Prisma create its connections

Ideally I think that fetching client options from the connector and using CloudRun created socket should work, but Prisma complains about the missing credentials

Something similar to:
```js
const clientOpts = connector.getOptions({
  instanceConnectionName
  authType: AuthTypes.IAM,
})

const datasourceUrl = `postgresql://${POSTGRES_USER}@localhost/${POSTGRES_DB}?host=/cloudsql/${instanceConnectionName}`

this.client = new PrismaClient({
  datasourceUrl,
  ...clientOpts,
})
```

**Comment by shanirosen-chram** [comment]
@Verhaeg Hi, thanks for the reply! thats exactly what we did with CloudRun and it worked, with k8s we used the same method but with a sidecar. 
The second option you mentioned did not work for us. 
We just hoped we could do everything through prisma and not use google's connectors and proxy.

**Comment by Verhaeg** [comment]
Our solution so far was to fallback to user+password for now, together with the cloudsql connector in cloudrun (socket connection).

It would be great to be able to access using the ServiceAccount as it doesn't depend on a password, but for now it seems it needs to create a second procy socket.

**Comment by phillipstack** [comment]
This issue occurrs for me as well. We have been using the postgresql://<USER>:<PASSWORD>@localhost/<DATABASE>?host=/cloudsql/<CONNECTION_INSTANCE_NAME> format. 

What we are finding is that it seems to append port 5432 to the url

We configure this

postgresql://<USER>:<PASSWORD>@localhost/<DATABASE>?host=/cloudsql/<CONNECTION_INSTANCE_NAME>

and we see this in the logs

Can't reach database server at `/cloudsql/<CLOUDSQL_INSTANCE_RESOURCE_NAME>`:`5432`

We believe the appending of the 5432 is related to the error and would prefer not to implement alternate solutions and network configurations. Does anyone have a solution for this not involving a custom prisma client invocation?

**Comment by plopezlpz** [comment]
same thing here as @phillipstack `prisma migrate` keeps appending the port and doesn't find the socket. Any way to fix this?

```
Error: P1001: Can't reach database server at `/cloudsql/PROJECT:REGION:DB:5432`"
```

Using prisma 7.3.0

**Comment by laser-beams** [comment]
I am encountering same issue like @plopezlpz 
```
Error: P1001: Can't reach database server at `/cloudsql/PROJECT:REGION:DB:5432`"
```
Using prisma 6.6.0

**Comment by phillipstack** [comment]
So, I think in the end, I was never able to resolve the socket issue. I ended up using a VPC connection rather than the socket approach. This allows you to use a standard URI to connect.
